fvm/src/blockstore/buffered.rs

```diff
@@ -62,86 +56,63 @@ where
 /// This was implemented because the CBOR library we use does not expose low
 /// methods like this, requiring us to deserialize the whole CBOR payload, which
 /// is unnecessary and quite inefficient for our usecase here.
-fn cbor_read_header_buf<B: Read>(br: &mut B, scratch: &mut [u8]) -> anyhow::Result<(u8, usize)> {
-    let first = br.read_u8()?;
+fn cbor_read_header_buf<B: Read>(br: &mut B) -> anyhow::Result<(u8, u64)> {
+    #[inline(always)]
+    pub fn read_fixed<const N: usize>(r: &mut impl Read) -> std::io::Result<[u8; N]> {
+        let mut buf = [0; N];
+        r.read_exact(&mut buf).map(|_| buf)
+    }
+
+    let first = read_fixed::<1>(br)?[0];
     let maj = (first & 0xe0) >> 5;
     let low = first & 0x1f;
 
-    if low < 24 {
-        Ok((maj, low as usize))
-    } else if low == 24 {
-        let val = br.read_u8()?;
-        if val < 24 {
-            return Err(anyhow!(
-                "cbor input was not canonical (lval 24 with value < 24)"
-            ));
-        }
-        Ok((maj, val as usize))
-    } else if low == 25 {
-        br.read_exact(&mut scratch[..2])?;
-        let val = BigEndian::read_u16(&scratch[..2]);
-        if val <= u8::MAX as u16 {
-            return Err(anyhow!(
-                "cbor input was not canonical (lval 25 with value <= MaxUint8)"
-            ));
-        }
-        Ok((maj, val as usize))
-    } else if low == 26 {
-        br.read_exact(&mut scratch[..4])?;
-        let val = BigEndian::read_u32(&scratch[..4]);
-        if val <= u16::MAX as u32 {
-            return Err(anyhow!(
-                "cbor input was not canonical (lval 26 with value <= MaxUint16)"
-            ));
-        }
-        Ok((maj, val as usize))
-    } else if low == 27 {
-        br.read_exact(&mut scratch[..8])?;
-        let val = BigEndian::read_u64(&scratch[..8]);
-        if val <= u32::MAX as u64 {
-            return Err(anyhow!(
-                "cbor input was not canonical (lval 27 with value <= MaxUint32)"
-            ));
-        }
-        Ok((maj, val as usize))
-    } else {
-        Err(anyhow!("invalid header cbor_read_header_buf"))
-    }
+    let val = match low {
+        ..=23 => low.into(),
+        24 => read_fixed::<1>(br)?[0].into(),
+        25 => u16::from_be_bytes(read_fixed(br)?).into(),
+        26 => u32::from_be_bytes(read_fixed(br)?).into(),
+        27 => u64::from_be_bytes(read_fixed(br)?),
+        _ => return Err(anyhow!("invalid header cbor_read_header_buf")),
+    };
```

Comment:
Great refactor for improving readability! 

Regarding performance, IIUC this is now doing 2 extra allocations for each call now instead of using a scratch buffer as before?

Reply:
Stack allocations so it should basically be free. Technically, it might cause some zeroing, but that shouldn't matter (and may not even happen if rust can optimize it).